rclcpp/test/rclcpp/wait_set_policies/test_static_storage.cpp

```diff
@@ -0,0 +1,195 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <memory>
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/wait_set.hpp"
+#include "../../utils/rclcpp_gtest_macros.hpp"
+
+#include "test_msgs/msg/empty.hpp"
+#include "test_msgs/srv/empty.hpp"
+
+class TestStaticStorage : public ::testing::Test
+{
+public:
+  static void SetUpTestCase()
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  static void TearDownTestCase()
+  {
+    rclcpp::shutdown();
+  }
+
+  void SetUp()
+  {
+    node = std::make_shared<rclcpp::Node>("node", "ns");
+  }
+
+  std::shared_ptr<rclcpp::Node> node;
+};
+
+class TestWaitable : public rclcpp::Waitable
+{
+public:
+  TestWaitable()
+  : is_ready_(false) {}
+  bool add_to_wait_set(rcl_wait_set_t *) override {return true;}
+
+  bool is_ready(rcl_wait_set_t *) override {return is_ready_;}
+
+  void execute() override {}
+
+  void set_is_ready(bool value) {is_ready_ = value;}
+
+private:
+  bool is_ready_;
+};
+
+TEST_F(TestStaticStorage, iterables_construct_destruct) {
+  auto subscription = node->create_subscription<test_msgs::msg::Empty>(
+    "topic", 10, [](test_msgs::msg::Empty::SharedPtr) {});
+  // This is long, so it can stick around and be removed
+  auto timer = node->create_wall_timer(std::chrono::seconds(100), []() {});
+  auto guard_condition = std::make_shared<rclcpp::GuardCondition>();
+  auto service =
+    node->create_service<test_msgs::srv::Empty>(
+    "service",
+    [](
+      const test_msgs::srv::Empty::Request::SharedPtr,
+      test_msgs::srv::Empty::Response::SharedPtr) {});
+  auto client = node->create_client<test_msgs::srv::Empty>("service");
+  auto waitable = std::make_shared<TestWaitable>();
+  rclcpp::StaticWaitSet<1, 1, 1, 1, 1, 1> wait_set(
+    {{{subscription}}}, {guard_condition}, {timer}, {client}, {service}, {{{waitable}}});
+
+  EXPECT_TRUE(rcl_wait_set_is_valid(&wait_set.get_rcl_wait_set()));
+}
+
+TEST_F(TestStaticStorage, fixed_storage_needs_pruning) {
+  {
+    using StaticWaitSet = rclcpp::StaticWaitSet<1, 0, 0, 0, 0, 0>;
```

Comment:
Why does those cases fail when making it a shared pointer?

Reply:
It has nothing to do with my use of shared pointer here. I am only forced to use that so that the constructor can be called in a single statement like this. For some reason, the compiler didn't like `RCLCPP_EXPECT_THROW_EQ({rclcpp::StaticWaitSet<1, 0, 0, 0, 0, 0> wait_set;}, std_runtime_error(...`, so I had to get a little more creative. However, it looks like Windows complains about the unused return value from `make_shared`, so in the new commit, I call `reset()` on it as well.

The reason the exception is thrown though, is due to an unusual situation where the statically declared size in the template (`<1, 0, ...>`) is larger than the arrays passed to the constructor. In this case with no arguments, the arrays are defaulted to `{}`, which default-initializes an array of non-zero size. Because the items added to the waitset are not valid, `storage_policy_common`'s `storage_rebuild_rcl_wait_set_with_sets` decides they need to be pruned, which shouldn't happen for static storage.

Reply:
Oh I see, mind to add a bit of this explanation below the test case name? otherwise it seems the problem is related to the shared pointer usage

Reply:
I added a bit more of a comment describing this situation